app/kuma-dp/pkg/dataplane/metrics/server.go

```diff
@@ -153,44 +172,150 @@ func rewriteMetricsURL(address string, port uint32, path string, queryModifier Q
 	ctx := req.Context()
 	out := make(chan []byte, len(s.applicationsToScrape))
+	contentTypes := make(chan expfmt.Format, len(s.applicationsToScrape))
 	var wg sync.WaitGroup
 	done := make(chan []byte)
 	wg.Add(len(s.applicationsToScrape))
 	go func() {
 		wg.Wait()
-		close(done)
 		close(out)
+		close(contentTypes)
+		close(done)
 	}()
 
 	for _, app := range s.applicationsToScrape {
 		go func(app ApplicationToScrape) {
 			defer wg.Done()
-			out <- s.getStats(ctx, req, app)
+			content, contentType := s.getStats(ctx, req, app)
+			out <- content
+
+			// It's possible to track the highest priority content type seen,
+			// but that would require mutex.
+			// I would prefer to calculate it later at one go
+			contentTypes <- contentType
 		}(app)
 	}
 
 	select {
 	case <-ctx.Done():
 		return
 	case <-done:
-		// default format returned by prometheus
-		writer.Header().Set("content-type", string(expfmt.FmtText))
-		for resp := range out {
-			if _, err := writer.Write(resp); err != nil {
-				logger.Error(err, "error while writing the response")
-			}
-			if _, err := writer.Write([]byte("\n")); err != nil {
-				logger.Error(err, "error while writing the response")
-			}
+		selectedCt := selectContentType(contentTypes, req.Header)
+		writer.Header().Set(hdrContentType, string(selectedCt))
+
+		// aggregate metrics of target applications and attempt to make them
+		// compatible with FmtOpenMetrics if it is the selected content type.
+		metrics := processMetrics(out, selectedCt)
+		if _, err := writer.Write(metrics); err != nil {
+			logger.Error(err, "error while writing the response")
+		}
+	}
+}
+
+func processMetrics(contents <-chan []byte, contentType expfmt.Format) []byte {
+	buf := new(bytes.Buffer)
+
+	for metrics := range contents {
+		// remove the EOF marker from the metrics, because we are
+		// merging multiple metrics into one response.
+		metrics = bytes.ReplaceAll(metrics, []byte("# EOF"), []byte(""))
+
+		if _, err := buf.Write(metrics); err != nil {
+			logger.Error(err, "error while writing the response to temporary buffer")
+		}
+		if _, err := buf.Write([]byte("\n")); err != nil {
+			logger.Error(err, "error while writing the response")
+		}
+	}
+
+	processedMetrics := processNewlineChars(buf.Bytes())
+	processedMetrics = append(processedMetrics, '\n')
+	buf.Reset()
+	_, err := buf.Write(processedMetrics)
+	if err != nil {
+		panic(err)
+	}
+
+	if contentType == expfmt.FmtOpenMetrics_1_0_0 || contentType == expfmt.FmtOpenMetrics_0_0_1 {
+		// make metrics OpenMetrics compliant
+		buf.Write([]byte("# EOF\n"))
+		return buf.Bytes()
+	}
+
+	return buf.Bytes()
+}
+
+// processNewlineChars takes byte data and returns a new byte slice
+// after trimming and deduplicating the newline characters.
+func processNewlineChars(byteData []byte) []byte {
+	return bytes.TrimSpace(dedup(byteData, '\n', '\n'))
+}
+
+func dedup(input []byte, old, new byte) []byte {
```

Comment:
`new` isn't used, let's just do the deduplication hardcoded with `\n` inline in  `processNewLineChars`

Reply:
In `dedup` when we see two consecutive chars matching `old`, we replace those two chars with `new`. But the algorithm is bit wrong. It only works as expected when `old == new`. So there is no point to give param `new`.
Anyway as we are not using this function anywhere else at this moment, as you suggested will do the deduplication inside `processNewlineChars` itself.